csharp/extractor/Semmle.Util/MemoizedFunc.cs

```diff
@@ -0,0 +1,48 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Concurrent;
+
+namespace Semmle.Util;
+
+public class MemoizedFunc<T1, T2> where T1 : notnull
+{
+    private readonly Func<T1, T2> f;
+    private readonly Dictionary<T1, T2> cache = new();
+
+    public MemoizedFunc(Func<T1, T2> f)
+    {
+        this.f = f;
+    }
+
+    public T2 this[T1 s]
```

Comment:
Nit: `MemoizedFunc<>` doesn't behave as a `Func<>`. It behaves as a dictionary. I would change this indexer to an `Invoke` method, and then the call-site would match the `Func<>` closer: 
```c#
var f0 = new Func<int, int>(i=>i);
var f1 = new MemoizedFunc<int, int>(f0);

f0.Invoke(1);
f1.Invoke(1);
```